dune/grid/common/WellConnections.cpp

```diff
@@ -0,0 +1,197 @@
+        Opm::CompletionSetConstPtr completionSet = well->getCompletions(last_time_step);
+        for (size_t c=0; c<completionSet->size(); c++) {
+            Opm::CompletionConstPtr completion = completionSet->get(c);
+            int i = completion->getI();
+            int j = completion->getJ();
+            int k = completion->getK();
+            int cart_grid_idx = i + cartesianSize[0]*(j + cartesianSize[1]*k);
+            int compressed_idx = cartesian_to_compressed[cart_grid_idx];
+            if ( compressed_idx >= 0 ) // Ignore completions in inactive cells.
+            {
+                well_indices.insert(compressed_idx);
+            }
+        }
+        ++index;
+    }
+}
+
+std::vector<std::vector<int> >
+postProcessPartitioningForWells(std::vector<int>& parts,
+                                const Opm::EclipseStateConstPtr eclipseState,
+                                const WellConnections& well_connections,
+                                std::size_t no_procs)
+{
+    std::vector<const Opm::Well*>  wells  = eclipseState->getSchedule()->getWells();
+    // Contains for each process the indices of the wells assigned to it.
+    std::vector<std::vector<int> > well_indices_on_proc(no_procs);
+
+    if( ! well_connections.size() )
+    {
+        // No wells to be processed
+        return well_indices_on_proc;
+    }
+
+    // prevent memory allocation
+    for(auto& well_indices : well_indices_on_proc)
+    {
+        well_indices.reserve(wells.size());
+    }
+
+    // Check that all completions of a well have ended up on one process.
+    // If that is not the case for well then move them manually to the
+    // process that already has the most completions on it.
+    int well_index = 0;
+
+    for (const auto* well: wells) {
+        const auto& well_completions = well_connections[well_index];
+        std::map<int,std::size_t> no_completions_on_proc;
+        for ( auto completion_index: well_completions )
+        {
+            ++no_completions_on_proc[parts[completion_index]];
+        }
+
+        int owner = no_completions_on_proc.begin()->first;
+
+        if ( no_completions_on_proc.size() > 1 )
+        {
+            // partition with the most completions on it becomes new owner
+            int new_owner = std::max_element(no_completions_on_proc.begin(),
+                                             no_completions_on_proc.end(),
+                                             [](const std::pair<int,std::size_t>& p1,
+                                                const std::pair<int,std::size_t>& p2){
+                                                 return ( p1.second > p2.second );
+                                             })->first;
+            std::cout << "Manually moving well " << well->name() << " to partition "
+                      << new_owner << std::endl;
+
+            for ( size_t c = 0; c < well_completions.size(); c++ )
+            {
+                parts[c] = new_owner;
```

Comment:
Is this correct? Should it not rather be `parts[well_completions[c]] = new_owner`? Might be simpler if using range-for as on line 101.